### ISSUE TITLE
[ttnn][matmul] 2D mcast with width-sharded DRAM in1 silently returns garbage when `per_core_N < ceil(N / num_dram_channels)`

### ISSUE THREAD

**Issue by amirnd51** [tenstorrent/tt-metal#57732]
**Component:** `ttnn/cpp/ttnn/operations/matmul/device/factory/matmul_multicore_reuse_mcast_2d_program_factory.cpp`
**Version:** tt-metal `v0.77.0` (`9f9cd4fd`)
**Hardware:** Blackhole P150 (13×10 compute grid, 8 DRAM channels)
**Severity:** silent wrong results — no error is raised

### Summary

`ttnn.linear` / `ttnn.matmul` with a `MatmulMultiCoreReuseMultiCastProgramConfig` and a weight
that is `WIDTH_SHARDED` in DRAM (the layout `tt_transformers` uses for every prefill linear)
produces non-finite output whenever the per-worker block is narrower than one DRAM shard, i.e.
`per_core_N < ceil(N / device->num_dram_channels())`. The config passes validation and runs.

### Reproduction

`tests/ttnn_regressions/test_dram_sharded_narrow_worker_block.py` (self-contained, one device):
Llama-3.1-8B fused QKV shape `[1,1,128,4096] @ [1,1,4096,6144]`, weight DRAM width-sharded
across the device's DRAM cores, `in0_block_w=8`, grid 4 rows × `grid_x` columns.

| case | per_core_N vs shard width (24 tiles) | v0.77.0 | with fix |
|---|---|---|---|
| grid 8 | 24 = 24 | pass | pass |
| grid 9 | 22 < 24 | **non-finite output** | pass |
| grid 13 | 15 < 24 | **non-finite output** | pass |

A predicted rule — correct iff `per_core_N ≥ ceil(N/8)` — held on 22 of 22 configs across four
real layer shapes (N = 128, 192, 448), including grid widths 5–7 not seen before prediction, and
on all 288 configs of a checked search (widths 10–13: 229/229 wrong; width 8: 59/59 correct).

### Root cause

The factory walks the DRAM shards to fill each worker's in1 block, with cursor state carried
across worker columns. The **first** read for each worker is not clamped to the worker block:

```cpp
worker_core_stride = per_core_N_storage - storage_core_stride;        // :1371 and :2883
```

while every subsequent read is:

```cpp
stride = std::min(stride, per_core_N);                                // :1399 and :2911
```

When `per_core_N < per_core_N_storage` the first read takes more tiles than the worker's block,
the bank cursor advances past tiles the next column needed, and every later column reads wrong
weights.

The sibling factory `matmul_multicore_reuse_mcast_dram_sharded_program_factory.cpp` (used for
decode) contains the same unclamped line (`:826`) but reaches it only after branching on
`if (per_core_N_in1_sender < per_core_N_storage)` (`:784`), whose path caps every read — so decode
is not affected, and the fix ports existing handling rather than adding new logic.

### Fix

```cpp
worker_core_stride = std::min(per_core_N_storage - storage_core_stride, per_core_N);
```

at both sites (`analysis/p8-clampfix.diff`, 8 insertions / 2 deletions). Alternatively, reject
`per_core_N < per_core_N_storage` in validation so the case errors instead of corrupting.

### Verification

- **Causation:** built in an isolated tree; the five configs that were garbage become correct
  (PCC ≥ 0.99988); a width-8 control is correct on both builds. The loaded `_ttnncpp.so` was
  verified from `/proc/self/maps`.
- **No regression on the touched path:** 212 existing tests (nightly `test_matmul_dram_sharded.py`,
  DRAM-sharded bias/activation, tiny-tile, linear and deepseek DRAM-sharded tests, plus a grid/linear
  control) give identical per-test outcomes on both builds: 188 passed, 24 skipped, 0 failed.
  None of them reaches the failing case, which is why it was not caught.
- **Not run:** the full test suite; Wormhole; P100 (7 DRAM channels).

### Impact

The bug confines `tt_transformers` prefill configs on a P150 to ≤ 8-column grids. With the fix,
13-column grids keeping the model's own `in0_block_w` give, on a full Llama-3.1-8B prefill
(dummy weights, batch 1):

| prefill tokens | shipped configs | 13-column grids | speedup | logits |
|---|---|---|---|---|
| 128 | 54.20 ms | 44.61 ms | 1.215× | bit-identical |
| 512 | 81.89 ms | 71.42 ms | 1.147× | bit-identical |
| 1024 | 147.73 ms | 129.25 ms | 1.143× | bit-identical |

(At 512/1024 only gate/up and output are widened; down and QKV use `MinimalMatmulConfig`.)

---

## Related observations (separate issues, lower confidence of being bugs)

**1. `tt_transformers` QKV prefill config marked FIXME.** `get_attn_qkv_program_config` at seq ≤ 128
returns `in0_block_w=1`, subblock 1×1 (`# FIXME: optimize this config for prefill`). On a P150,
with the model's sharded weights, a correct `g8×4 b8 s1×4` config is 1.85× faster on device
(221.6 → 119.6 µs) at the same grid width, i.e. without the fix above. Swapping only this config
in a real decoder layer: 1.018× [1.005, 1.031].

**2. Auto program config degenerates to `in0_block_w=1` on Blackhole.** The all-DRAM 2D path
(`matmul_program_config.cpp:1301`) sets `in0_block_w = Kt % num_cores_x == 0 ? Kt / num_cores_x : 1`
with `num_cores_x` from `compute_with_storage_grid_size()` = 13 on a P150. No power-of-two tile
count is divisible by 13, so the result is always 1 there; on an 8-wide grid the same expression
yields hand-tuned values. For square bfp8_b matmuls this costs up to 1.98× on device versus the
same config with `in0_block_w=4` (identity confirmed by JIT cache key at 1024³). Only reached when no
program config is passed. The Wormhole side of this is inferred from source and untested.

---

---

Found while characterising Blackhole P150 for an academic study of the
Tenstorrent stack. Happy to provide additional detail or re-run anything on request.
